### DISCUSSION TITLE
atomsWithQuery: the component didn't re-render when mutating data from another Component

### DISCUSSION THREAD

**Discussion by meotimdihia** [pmndrs/jotai#1608] [Q&A] [answered]
Even though the state is changed, but component didn't re-render. It is very weird, This is an example: 

https://stackblitz.com/edit/next-typescript-pokptk-tkdiff?file=pages%2Findex.tsx

https://user-images.githubusercontent.com/300961/204875203-43ffc5ef-384d-4d00-809d-04374e9873d5.mov

**Comment by milang** [comment, marked as answer]
@meotimdihia your issue seems to be related to referential equality.

If you replace your in-place object mutation on lines 63-69 in your prototype:
```ts
onSuccess: (result, data) => {
  queryClient.setQueryData(['user'], (oldData: any) => {
    oldData.username = data;
    return oldData;
  });
},
```

With a new object:
```ts
onSuccess: (_, data) => {
  queryClient.setQueryData(['user'], { username: data });
},
```

Both buttons then cause the username text to be updated.

Your approach keeps the `['user']` cache item pointing to the same object (same reference), so unless observers perform deep equality check they won't realize there was a change.

PS: If you have other properties in the data object, you can use the spread operator to preserve them:

```ts
onSuccess: (_, data) => {
  queryClient.setQueryData(['user'], (oldData: any) => ({
    ...oldData,
    username: data,
  }));
},
```

**Reply by meotimdihia** [reply]
Thanks for the explanation.
